doc/tutorials/code-generator.md

```diff
@@ -0,0 +1,240 @@
+
+### Creating the supporting class
+
+Since we need to some way of supplying data to our template, we have to create a subclass of `GenTemplate`.
+
+First we create a new directory for our template class in `src/tools/kdb/gen`, for our template it should be `src/tools/kdb/gen/example`. In
+this directory we then create `example.cpp` and `example.hpp`. If your template class becomes sufficiently complex, it may make sense to
+split the code into multiple classes and into multiple files, for this reason we recommend creating a new directory for each template.
+The CMake script will also automatically recognize your files, if you put them directly into `src/tools/kdb/gen`, but using additional
+subdirectories (beyond the one matching your template name) like `src/tools/kdb/gen/example/src` would require modifying
+[`src/tools/kdb/CMakeLists.txt`](/src/tools/kdb/CMakeLists.txt).
+
+In `example.cpp` and `example.hpp` we create our subclass of `GenTemplate`. Therefore `example.hpp` should look like this:
+
+```cpp
+#ifndef ELEKTRA_EXAMPLE_HPP // choose a unique header guard
+#define ELEKTRA_EXAMPLE_HPP
+
+#include <gen/template.hpp>
+
+class ExampleGenTemplate : public GenTemplate
+{
+public:
+    ExampleGenTemplate () : GenTemplate ("example", { ".txt" }, {}, {})
+    {
+    }
+
+protected:
+    kainjow::mustache::data getTemplateData (const std::string & outputName, const std::string & part, const kdb::KeySet & ks,
+                                             const std::string & parentKey) const override;
+};
+
+#endif // ELEKTRA_EXAMPLE_HPP
+```
+
+Apart from the line `ExampleGenTemplate () : GenTemplate ("example", { ".txt" }, {}, {})` everything should be more or less the same for
+all templates. Let's dissect this line:
+
+- `ExampleGenTemplate ()` we declare a zero argument constructor. All templates _must_ have only a single zero argument constructor,
+  otherwise the framework cannot instantiate them.
+- `: GenTemplate (` the constructor must invoke the base-class constructor.
+- `"example",` we decided to use `example` as the base name for our template files. This base name will be replaced by the `outputName`
+  chosen by the user, when invoking the code-generator.
+- `{ ".txt" },` this is the list of part suffixes for our template. We only have a single part (output file) with the suffix `.txt`.
+- `{},` the first empty list would contain all the partials our template uses. We don't use any.
+- `{},` the second empty list, is actually a map. It would contain all parameters and whether they are required or not.
+  We don't have parameters.
+
+As you can see, the header is quite simple. The more important part is actually the source file. It contains the implementation of
+`getTemplateData`, the function that delivers all the template data to the framework.
+
+For our example we will use this implementation:
+
+```cpp
+kainjow::mustache::data ExampleGenTemplate::getTemplateData (const std::string & outputName, const std::string & part,
```

Comment:
The warnings are to be expected. I don't think it is the aim of this guide to explain how to avoid these warnings. I also think it is not wise to add `ELEKTRA_UNUSED` to the code in the example. If we did, people probably would forget to remove them, if the do use the parameter.

Reply:
Yes, the warnings is a minor problem, the header file is important.